extmod/moduevent_native.c

```diff
@@ -0,0 +1,244 @@
+/*
+ * This file is part of the MicroPython project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Damien P. George
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/runtime.h"
+#include "py/stream.h"
+#include "py/mphal.h"
+#include "extmod/moduevent.h"
+
+#if MICROPY_PY_UEVENT == MICROPY_PY_UEVENT_IMPL_NATIVE
+
+struct _mp_obj_poll_t {
+    mp_obj_base_t base;
+    unsigned short alloc;
+    unsigned short len;
+    struct _mp_obj_poll_entry_t *entries;
+    struct _mp_obj_poll_entry_t *entry_ready;
+    struct _mp_obj_poll_entry_t *entry_free;
+};
+
+typedef struct _mp_obj_poll_entry_t {
+    mp_obj_poll_entry_base_t base;
+    struct _mp_obj_poll_entry_t *next;
+    mp_uint_t (*ioctl)(mp_obj_t obj, mp_uint_t request, uintptr_t arg, int *errcode);
+    uint16_t flags_waiting;
+    uint16_t flags_ready;
+} mp_obj_poll_entry_t;
+
+STATIC mp_obj_poll_entry_t *mp_uevent_poll_entry_new(mp_obj_poll_t *poller) {
+    if (poller->len >= poller->alloc) {
+        size_t new_alloc = poller->alloc + MICROPY_ALLOC_UEVENT_POLL_INC;
+        poller->entries = m_renew(mp_obj_poll_entry_t, poller->entries, poller->alloc, new_alloc);
+        poller->alloc = new_alloc;
+    }
+    mp_obj_poll_entry_t *self = &poller->entries[poller->len++];
+    self->base.base.type = &mp_type_poll_entry;
+    self->base.poller = poller;
+    self->base.user_method_name = MP_QSTR_NULL;
+    self->base.user_data = mp_const_none;
+    self->next = NULL;
+    return self;
+}
+
+STATIC void mp_uevent_poll_entry_init(mp_obj_poll_entry_t *self, mp_obj_t obj) {
+    self->base.obj = obj;
+    self->base.user_flags = 0;
+    self->ioctl = mp_get_stream_raise(obj, MP_STREAM_OP_IOCTL)->ioctl;
```

Comment:
You're right, it doesn't need to be coupled with streams and it shouldn't be.  But I was trying to make a smooth transition from the current behaviour and not add unnecessary things.  So what it needs is for there to be a general ioctl protocol in MicroPython which is more fundamental than the stream protocol (it could be called an event protocol but I think it's fair enough to combine it with an ioctl function to minimise code).